### ISSUE TITLE
Compiled model retracing calls to random functions

### ISSUE THREAD

**Issue by AakashKumarNain** [keras-team/keras-core#762]
I was debugging some code when I noticed that layers utilizing random functions in the backend in a `jit` compiled model are being retraced at every call with TensorFlow and JAX backend. I added a print statement in the `dropout` function in the backend. Here is the code to reproduce it

```python
def build_model():
    inp = layers.Input(shape=(3,), batch_size=1)
    out = layers.Dropout(0.5, seed=1234)(inp)
    return Model(inp, out)

model = build_model()
model.compile(jit_compile=True)

# Test the model for a constant input
x = ops.ones((1, 3))
drop_layer = model.layers[-1]
for _ in range(5):
    y = model(x, training=True)
    print(" | Generator state: ", drop_layer.seed_generator.state.value)
    print("Model output: ", y)
    print("="*75)
```
### Output with TF backend
```
Dropout Called  | Generator state:  [1234    1]
Model output:  tf.Tensor([[2. 2. 0.]], shape=(1, 3), dtype=float32)
===========================================================================

Dropout Called  | Generator state:  [1234    2]
Model output:  tf.Tensor([[2. 0. 0.]], shape=(1, 3), dtype=float32)
===========================================================================

Dropout Called  | Generator state:  [1234    3]
Model output:  tf.Tensor([[0. 2. 2.]], shape=(1, 3), dtype=float32)
===========================================================================

Dropout Called  | Generator state:  [1234    4]
Model output:  tf.Tensor([[0. 2. 2.]], shape=(1, 3), dtype=float32)
===========================================================================

Dropout Called  | Generator state:  [1234    5]
Model output:  tf.Tensor([[2. 0. 2.]], shape=(1, 3), dtype=float32)
===========================================================================
```

### Output with JAX backend
```
Dropout called  | Generator state:  [1234    1]
Model output:  [[2. 2. 2.]]
===========================================================================

Dropout called  | Generator state:  [1234    2]
Model output:  [[0. 2. 2.]]
===========================================================================

Dropout called  | Generator state:  [1234    3]
Model output:  [[2. 2. 2.]]
===========================================================================

Dropout called  | Generator state:  [1234    4]
Model output:  [[2. 0. 2.]]
===========================================================================

Dropout called  | Generator state:  [1234    5]
Model output:  [[2. 2. 2.]]
===========================================================================
```

# Expected output
Irrespective of the backend, the random functions once compiled shouldn't be retraced unless the shape or dtype of the input arguments to the function changes

**Comment by fchollet** [comment]
You are running eagerly when you are doing `y = model(x, training=True)`. `. compile()` does not affect `__call__`.

The compiled functions are only used through `model.fit()`, `model.evaluate()`, `model.predict()`, and their single batch equivalents (e.g. `model.train_on_batch()`). Otherwise if you want a compiled function you should define your own via `jax.jit` or `tf.function`.

**Comment by AakashKumarNain** [comment]
Oh okay. I thought `run_eagerly=False` was taking care of that. Thanks for the clarification @fchollet
